presto-native-execution/presto_cpp/main/runtime-metrics/prometheus-reporter/PrometheusReporter.cpp

```diff
@@ -0,0 +1,231 @@
+
+#include <prometheus/collectable.h>
+#include <prometheus/counter.h>
+#include <prometheus/gauge.h>
+#include <prometheus/histogram.h>
+#include <prometheus/registry.h>
+#include <prometheus/summary.h>
+#include <prometheus/text_serializer.h>
+
+namespace facebook::presto::prometheus {
+
+using CounterMap = folly::Synchronized<
+    std::unordered_map<std::string, ::prometheus::Counter&>>;
+using GaugeMap =
+    folly::Synchronized<std::unordered_map<std::string, ::prometheus::Gauge&>>;
+using HistogramMap = folly::Synchronized<
+    std::unordered_map<std::string, ::prometheus::Histogram&>>;
+using SummaryMap = folly::Synchronized<
+    std::unordered_map<std::string, ::prometheus::Summary&>>;
+
+static constexpr std::string_view kSummarySuffix("_summary");
+using namespace ::prometheus;
+
+struct PrometheusReporter::Impl {
+  Impl(const Labels& labels) {
+    registry = std::make_shared<::prometheus::Registry>();
+    for (auto& itr : labels) {
+      this->labels[itr.first] = itr.second;
+    }
+  }
+
+ public:
+  ::prometheus::Labels labels;
+  std::shared_ptr<::prometheus::Registry> registry;
+  // Map of metric name and Metric type object (Counter, Gauge etc.).
+  // The prometheus::Registry holds these objects as well, but, doesn't provide
+  // an interface to fetch metric object by name. We maintain this
+  // map, so we can update a metric value when addMetric() is called.
+  CounterMap countersMap;
+  GaugeMap gaugeMap;
+  HistogramMap histogramMap;
+  SummaryMap summaryMap;
+};
+
+PrometheusReporter::PrometheusReporter(
+    const std::map<std::string, std::string>& labels) {
+  impl_ = std::make_shared<Impl>(labels);
+}
+void PrometheusReporter::registerMetricExportType(
+    const char* key,
+    facebook::velox::StatType statType) const {
+  if (registeredMetrics_.count(key)) {
+    // Already registered;
+    VLOG(1) << "Trying to register already registered metric " << key;
+    return;
+  }
+  // Prometheus format requires '.' to be replaced.
+  std::string displayKey = std::string(key);
+  std::replace(displayKey.begin(), displayKey.end(), '.', '_');
+  switch (statType) {
+    case facebook::velox::StatType::COUNT: {
+      auto& counterFamily =
+          BuildCounter().Name(displayKey).Register(*impl_->registry);
+      auto& counter = counterFamily.Add(impl_->labels);
+      impl_->countersMap.wlock()->emplace(std::string(key), counter);
+      break;
+    }
+    case facebook::velox::StatType::SUM:
+    case facebook::velox::StatType::AVG:
+    case facebook::velox::StatType::RATE: {
+      auto& gaugeFamily =
+          BuildGauge().Name(displayKey).Register(*impl_->registry);
+      auto& gauge = gaugeFamily.Add(impl_->labels);
+      impl_->gaugeMap.wlock()->emplace(std::string(key), gauge);
+      break;
+    }
+    default:
+      VELOX_UNSUPPORTED(
+          "Unsupported metric type {}", std::to_string((int)statType));
+  }
+  registeredMetrics_.emplace(key, statType);
```

Comment:
I want to avoid the double lookup for each metric (in `registeredMetrics_` and `gaugeMaps` etc.).
How about we store a pair `<statType, void*>` as the value for `registeredMetrics_`.
`void*` points to the address of the counter returned. We can type-cast when we update and get each metric.